internal/cache/file/cache_handler.go

```diff
@@ -0,0 +1,181 @@
+	// fileInfoCache contains the reference of fileInfo cache.
+	fileInfoCache *lru.Cache
+
+	// jobManager contains reference to a singleton jobManager.
+	jobManager *downloader.JobManager
+
+	// Directory location which would contain local file to keep the downloaded data.
+	cacheLocation string
+
+	// Guards the handling of cache entry insertion while creating the CacheHandle.
+	// Also guards post eviction logic E.g. cancelling and deletion of async download job,
+	// deletion of local file containing downloaded data.
+	mu locker.Locker
+}
+
+func NewCacheHandler(fileInfoCache *lru.Cache, jobManager *downloader.JobManager, cacheLocation string) *CacheHandler {
+	return &CacheHandler{
+		fileInfoCache: fileInfoCache,
+		jobManager:    jobManager,
+		cacheLocation: cacheLocation,
+		mu:            locker.New("FileCacheHandler", func() {}),
+	}
+}
+
+func (chr *CacheHandler) createLocalFileReadHandle(objectName string, bucketName string) (*os.File, error) {
+	fileSpec := data.FileSpec{
+		Path: util.GetDownloadPath(chr.cacheLocation, util.GetObjectPath(bucketName, objectName)),
+		Perm: util.DefaultFilePerm,
+	}
+
+	return util.CreateFile(fileSpec, os.O_RDONLY)
+}
+
+func (chr *CacheHandler) performPostEvictionWork(fileInfo *data.FileInfo) error {
+	key := fileInfo.Key
+	_, err := key.Key()
+	if err != nil {
+		return fmt.Errorf("error while performing post eviction: %v", err)
+	}
+
+	chr.jobManager.RemoveJob(key.ObjectName, key.BucketName)
+
+	localFilePath := util.GetDownloadPath(chr.cacheLocation, util.GetObjectPath(key.BucketName, key.ObjectName))
+	err = os.Remove(localFilePath)
+	if err != nil {
+		return fmt.Errorf("while deleting file: %s, error: %v", localFilePath, err)
+	}
+
+	return nil
+}
+
+// Acquires and releases LOCK(CacheHandler.mu)
+func (chr *CacheHandler) addFileInfoEntryInTheCacheIfNotAlready(object *gcs.MinObject, bucket gcs.Bucket) error {
+	fileInfoKey := data.FileInfoKey{
+		BucketName: bucket.Name(),
+		ObjectName: object.Name,
+	}
+	fileInfoKeyName, err := fileInfoKey.Key()
+	if err != nil {
+		return fmt.Errorf("while creating key: %v", fileInfoKeyName)
+	}
+
+	chr.mu.Lock()
+	defer chr.mu.Unlock()
+
+	// Todo - (raj-prince) - this lookUp should not change the LRU order.
+	fileInfo := chr.fileInfoCache.LookUp(fileInfoKeyName)
+	if fileInfo == nil {
+		fileInfo = data.FileInfo{
+			Key:              fileInfoKey,
+			ObjectGeneration: object.Generation,
+			Offset:           0,
+			FileSize:         object.Size,
+		}
+
+		evictedValues, err := chr.fileInfoCache.Insert(fileInfoKeyName, fileInfo)
+		if err != nil {
+			return fmt.Errorf("while inserting into the cache: %v", err)
+		}
+
+		for _, val := range evictedValues {
+			fileInfo := val.(data.FileInfo)
+			err := chr.performPostEvictionWork(&fileInfo)
+			if err != nil {
+				return fmt.Errorf("while performing post eviction of %s object error: %v", fileInfo.Key.ObjectName, err)
+			}
+		}
+	} else {
+		job := chr.jobManager.GetJob(object, bucket)
```

Comment:
I don't think this check is required here. Although we can add check which will act like as an assert that - fileInfo.generation will always be equal to object.generation.

The reason is for every kernel read fileHandle's tryEnsure is called - 

``` go
// If we already have a reader, and it's at the appropriate generation, we
	// can use it. Otherwise we must throw it away.
	if fh.reader != nil {
		if fh.reader.Object().Generation == fh.inode.SourceGeneration().Object {
			return
		}

		fh.reader.Destroy()
		fh.reader = nil
	}
```

So, we will invalidate the cache as part of destroy method only.

Reply:
I think it should be required in the below case
1. Open, Read  and close file A - File A is cached with Gen 1.
2. File A is changed remotely - The backing Object's Gen becomes 2.
3. Stat cache expired
5. Open, Read and close file A again - The above check that you mentioned will not be triggered, right ? (because at the time of opening file, LookUpInode will be called and that will update the object's generation to 2). So, we will end up reading stale data from file in cache ?

Reply:
Got it. In this case, object.generation == inode.generation but greater than fileInfo.generation.